cupy/cuda/__init__.py

```diff
@@ -53,13 +54,15 @@ def __init__(self, name):
 
 try:
     from cupy.cuda import nccl  # NOQA
-except ImportError:
+except ImportError as e:
```

Comment:
To support preloading we need to separate this code into another file like done in `cudnn.py`, and remove import from `__init__.py`. Otherwise, warnings will be shown every time running `import cupy` when nccl/cutensor is unavailable.

Strictly speaking this is backward incompatible as this breaks user code (e.g., `import cupy.cuda; print(cupy.cuda.nccl)`), but I think it's acceptable.

Reply:
You are right, I was thinking to always warn if libs are not found at CuPy import time, but perhaps it's too much noise.

@kmaehashi in this case, let me send a separate PR moving nccl to cupy_backends (which is something long overdue anyway), and then add cupy/cuda/nccl.py to enable this.

Reply:
Separate PR sounds good, thank you!

Reply:
See #4919. @kmaehashi I suppose we also need a `cupy/cuda/cutensor.py`?

Reply:
Yes!

Reply:
> I suppose we also need a `cupy/cuda/cutensor.py`?

See #4920.

Reply:
Both #4919 and #4920 are in!